src/Multi_persistence/include/gudhi/Multi_persistence/Line.h

```diff
@@ -5,253 +5,267 @@
-   */
-  template <typename U = T>
-  inline U push_forward2(const point_type &x) const;
-  /*
-   * Retuns the time parameter of the coordinate given by push_forward.
-   */
-  template <typename U = T>
-  inline U push_forward2(const kcritical_point_type &x) const;
-  /*
-   * Returns the point whose intersection is \f$ \max\{ y\le x \} \cap \mathrm{this}\f$
+  using K_critical_point = Gudhi::multi_filtration::Multi_critical_filtration<T>;
+
+  /**
+   * @brief Default constructor. Sets the number of coordinates to 0.
    */
-  inline point_type push_back(point_type x) const;
-  /*
-   * Retuns the time parameter of the coordinate given by push_back.
+  Line() : basePoint_(0), direction_(0) {}  // has to be explicitly set to 0, otherwise becomes -inf
+  /**
+   * @brief Constructs a line going through the given point with slope 1.
+   *
+   * @param x A point of the line.
    */
-  template <typename U = T>
-  inline U push_back2(const point_type &x) const;
-  /*
-   * Retuns the time parameter of the coordinate given by push_back.
+  Line(const Point &x) : basePoint_(x), direction_(0) {}  // default direction
+  /**
+   * @brief Constructs a line going through the given point with slope 1.
+   *
+   * @param x A point of the line. Will be moved.
    */
-  template <typename U = T>
-  inline U push_back2(const kcritical_point_type &x) const;
-  inline int get_dim() const;
-  /*
-   * Given a box, returns the coordinates of the intersection of this box and `this` as a pair of points (low, high)
-   * in this line, representing this interval.
+  Line(Point &&x) : basePoint_(std::move(x)), direction_(0) {}  // default direction
+  /**
+   * @brief Constructs a line going through the given point in the direction of the given vector.
+   * If the vector has no coordinates, the slope is assumed to be 1.
+   * Otherwise, the vector has to be non trivial and all its coordinates have to be positive.
+   *
+   * @param x A point of the line.
+   * @param vector Direction of the line. Positive and non trivial.
    */
-  std::pair<point_type, point_type> get_bounds(const Box<T> &box) const;
-  /*
-   * Retuns the times parameter of the coordinates in the pair given by get_bounds.
+  Line(const Point &x, const Point &vector) : basePoint_(x), direction_(vector) { check_direction_(); }
+
+  /**
+   * @brief Returns the coordinates of the point on the line with "time" parameter `t`. That is, the point \f$ x \f$
+   * such that \f$ x[i] = base\_point[i] + t \times direction[i] \f$ for all \f$ i \in [0, n - 1] \f$ with \f$ n \f$
+   * the number of coordinates.
    */
-  std::pair<T, T> get_bounds2(const Box<T> &box) const;
+  Point operator[](T t) const {
+    Point x(basePoint_.size());
+
+    if (direction_.size() > 0) {
+      for (std::size_t i = 0; i < x.size(); i++) x[i] = basePoint_[i] + t * direction_[i];
+    } else
+      for (std::size_t i = 0; i < x.size(); i++) x[i] = basePoint_[i] + t;
```

Comment:
This can segfault if direction is of size 0<s<x.size(). IIRC, my convention was to assume it was 1 if not set, although this might be confusing. I'm not using this convention, so this is fine by me anyway. We can throw before the beginning of the loop to be safe instead.